admin/provision.go

```diff
@@ -0,0 +1,143 @@
+package admin
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/google/uuid"
+	"github.com/rilldata/rill/admin/database"
+	"github.com/rilldata/rill/admin/provisioner"
+)
+
+type ProvisionOptions struct {
+	DeploymentID string
+	Type         provisioner.ResourceType
+	Name         string
+	Provisioner  string
+	Args         map[string]any
+	Annotations  map[string]string
+}
+
+func (s *Service) Provision(ctx context.Context, opts *ProvisionOptions) (*database.ProvisionerResource, error) {
```

Comment:
In general if provisioning fails then deployment should be deleted ?

Reply:
Not anymore – if provisioning fails, the deployment will still be in Postgres with an error status. Provisioning can then be retried by doing a reset (the reset will delete the old deployment record in Postgres, so no leaks).

This enables moving provisioning to a retryable background job, where the UI/CLI can poll the deployment until its status becomes "ok". Kasper is planning to do this move when he gets some time.